features/include/pcl/features/impl/multiscale_feature_persistence.hpp

```diff
@@ -136,20 +137,24 @@ template <typename PointSource, typename PointFeature> void
 pcl::MultiscaleFeaturePersistence<PointSource, PointFeature>::calculateMeanFeature ()
 {
   // Reset mean feature
-  for (int i = 0; i < feature_representation_->getNumberOfDimensions (); ++i)
-    mean_feature_[i] = 0.0f;
+  std::fill_n(mean_feature_.begin (), mean_feature_.size (), 0.f);
 
-  float normalization_factor = 0.0f;
+  std::size_t normalization_factor = 0;
   for (const auto& scale: features_at_scale_vectorized_)
   {
-    normalization_factor += static_cast<float> (scale.size ());
+    normalization_factor += scale.size ();  // not using accumulate for cache efficiency
     for (const auto &feature : scale)
-      for (int dim_i = 0; dim_i < feature_representation_->getNumberOfDimensions (); ++dim_i)
-        mean_feature_[dim_i] += feature[dim_i];
+      std::transform(mean_feature_.cbegin (), mean_feature_.cend (),
+                     feature.cbegin (), mean_feature_.begin (), std::plus<>{});
   }
 
-  for (int dim_i = 0; dim_i < feature_representation_->getNumberOfDimensions (); ++dim_i)
-    mean_feature_[dim_i] /= normalization_factor;
+  float factor = std::min<float>(1, normalization_factor);
+  std::transform(mean_feature_.cbegin(),
+                 mean_feature_.cend(),
+                 mean_feature_.begin(),
+                 [&factor](const auto& mean) {
```

Comment:
The usual advice goes that primitive types should be passed/captured by value. I was wondering what's your opinion on this.

Reply:
The advice is correct, specially since I'm not modifying it.

Reply:
I had both `factor` and `mean` in mind, by the way.

Reply:
About `mean`, it might be better to use reference. I've no strong feelings, since the body is small enough to be optimized in either case. The sole reason I used that is because I almost always go for `const auto&` in the lambda input.